### PR TITLE
Stabilize `maybe_uninit_ref`

### LINKED ISSUE #63568: Tracking issue for `#![feature(maybe_uninit_ref)]`
This is a tracking issue for the RFC "Deprecate `uninitialized` in favor of a new `MaybeUninit` type" (rust-lang/rfcs#1892).

This issue specifically tracks the following unstable methods:

```rust
impl<T> MaybeUninit<T> {
    pub unsafe fn assume_init_ref(&self) -> &T { ... }
    pub unsafe fn assume_init_mut(&mut self) -> &mut T { ... }
}
```
(NOTE: on current nightly, these methods are still called `get_ref` and `get_mut`, but the FCP here decided a rename.)

### PR DESCRIPTION
**Pull request by JohnTitor**
This stabilizes `assume_init_{ref,mut}`. FCP is complete: https://github.com/rust-lang/rust/issues/63568#issuecomment-590121300
The renaming was done by #76047 and FIXME was resolved by #76241, so I think we can now stabilize them finally 🎉 
Still, it's const-unstable as `assert_inhabited` is unstable.

Closes #63568

### CODE REVIEW

**Comment by JohnTitor** [comment]
cc @RalfJung

**Comment by JohnTitor** [comment]
Btw, should we also rename them on https://rust-lang.github.io/rfcs/1892-uninitialized-uninhabited.html#detailed-design?

File: library/core/src/mem/maybe_uninit.rs [hunk 1]
```diff
@@ -703,7 +700,7 @@ impl<T> MaybeUninit<T> {
     ///    // Reference to an uninitialized `Cell<bool>`: UB!
     /// }
     /// ```
-    #[unstable(feature = "maybe_uninit_ref", issue = "63568")]
+    #[stable(feature = "maybe_uninit_ref", since = "1.55.0")]
```
Comment (RalfJung) [inline library/core/src/mem/maybe_uninit.rs:703]: Shouldn't this be 1.54?

Comment (RalfJung) [reply library/core/src/mem/maybe_uninit.rs:703]: OTOH, it's release week, so this will probably land after the beta branches.

File: library/core/src/mem/maybe_uninit.rs [hunk 2]
```diff
@@ -823,10 +816,7 @@ impl<T> MaybeUninit<T> {
     ///     foo.assume_init()
     /// };
     /// ```
-    // FIXME(#76092): We currently rely on the above being incorrect, i.e., we have references
-    // to uninitialized data (e.g., in `libcore/fmt/float.rs`).  We should make
```
Comment (RalfJung) [inline library/core/src/mem/maybe_uninit.rs:827]: Why did you remove this FIXME? I don't think it has been resolved, has it?

File: library/core/src/mem/maybe_uninit.rs (hunk 1 above)
Comment (JohnTitor) [reply library/core/src/mem/maybe_uninit.rs:703]: #86234 landed already so it should be 1.55.0 IIUC.

File: library/core/src/mem/maybe_uninit.rs (hunk 2 above)
Comment (RalfJung) [reply library/core/src/mem/maybe_uninit.rs:827]: https://github.com/rust-lang/rust/pull/76241 fixed this for `fmt/float.rs`, but before removing the FIXME we need an audit of all the uses of these functions in the standard library. Note the "e.g.", this was not an exhaustive list.

**Comment by RalfJung** [comment]
> Btw, should we also rename them on https://rust-lang.github.io/rfcs/1892-uninitialized-uninhabited.html#detailed-design?

We don't usually update RFCs. They reflect the consensus at the time, they are not documentation of the current state of affairs.

File: library/core/src/mem/maybe_uninit.rs (hunk 2 above)
Comment (JohnTitor) [reply library/core/src/mem/maybe_uninit.rs:827]: Oh I thought it was fixed as the issue was closed. Then should we re-open it or have a new issue? And I still don't understand this very well, so if you could explain it to me, that would be great.

Comment (JohnTitor) [reply library/core/src/mem/maybe_uninit.rs:827]: Okay I greped the code and this is the only piece I found, other uses have debug assertions or safety comments:
https://github.com/rust-lang/rust/blob/14831568d506b5ee7be3e7d5a2f029ce9048b609/library/std/src/io/copy.rs#L128-L156

Can this be a blocker to be stabilized?

Comment (RalfJung) [reply library/core/src/mem/maybe_uninit.rs:827]: I agree the other uses are fine, and this code will hopefully disappear when https://github.com/rust-lang/rfcs/blob/master/text/2930-read-buf.md gets implemented.

So, fine to remove the FIXME. :)

**Comment by RalfJung** [comment]
r? @RalfJung 
Thanks. :)
Considering that this has been FCP'd more than a year ago, I think I can just approve this. Still Cc @rust-lang/libs 

@bors r+
